python/ql/src/semmle/python/frameworks/Stdlib.qll

```diff
@@ -864,6 +864,222 @@ private module Stdlib {
+      returnsPath.getAttributeName() = pathlibPathInjection() and
+      returnsPath.getObject().getALocalSource() = pathlibPath(t2) and
+      t2.end()
+    |
+      t.start() and
+      result.(DataFlow::CallCfgNode).getFunction() = returnsPath
+    )
+    or
+    // Track further
+    exists(DataFlow::TypeTracker t2 | result = pathlibPath(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `pathlib.Path` object. */
+  DataFlow::LocalSourceNode pathlibPath() { result = pathlibPath(DataFlow::TypeTracker::end()) }
+
+  private class PathlibFileAccess extends FileSystemAccess::Range, DataFlow::CallCfgNode {
+    DataFlow::AttrRead fileAccess;
+
+    PathlibFileAccess() {
+      fileAccess.getAttributeName() in [
+          "stat", "chmod", "exists", "expanduser", "glob", "group", "is_dir", "is_file", "is_mount",
+          "is_symlink", "is_socket", "is_fifo", "is_block_device", "is_char_device", "iter_dir",
+          "lchmod", "lstat", "mkdir", "open", "owner", "read_bytes", "read_text", "readlink",
+          "rename", "replace", "resolve", "rglob", "rmdir", "samefile", "symlink_to", "touch",
+          "unlink", "link_to", "write_bytes", "write_text"
+        ] and
+      fileAccess.getObject().getALocalSource() = pathlibPath() and
+      this.getFunction() = fileAccess
+    }
+
+    override DataFlow::Node getAPathArgument() { result = fileAccess.getObject() }
+  }
+
+  /** An additional taint steps for objects of type `pathlib.Path` */
+  private class PathlibPathTaintStep extends TaintTracking::AdditionalTaintStep {
+    override predicate step(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+      // Type construction
+      nodeTo = pathlib().getMember(pathlibPathConstructor()).getACall() and
+      nodeFrom = nodeTo.(DataFlow::CallCfgNode).getArg(_)
+      or
+      // Type preservation
+      nodeFrom.getALocalSource() = pathlibPath() and
+      nodeTo.getALocalSource() = pathlibPath() and
+      (
+        // Type-preserving call
+        typePreservingCall(nodeFrom, nodeTo)
+        or
+        // Type-preserving attribute
+        typePreservingAttribute(nodeFrom, nodeTo)
+      )
+      or
+      // Data injection
+      nodeTo.getALocalSource() = pathlibPath() and
+      (
+        // Special handling of the `/` operator
+        exists(BinaryExprNode slash, DataFlow::Node pathOperand, DataFlow::Node dataOperand |
+          slash.getOp() instanceof Div and
+          (
+            pathOperand.asCfgNode() = slash.getLeft() and
+            dataOperand.asCfgNode() = slash.getRight()
+            or
+            pathOperand.asCfgNode() = slash.getRight() and
+            dataOperand.asCfgNode() = slash.getLeft()
+          ) and
+          pathOperand.getALocalSource() = pathlibPath()
+        |
+          nodeTo.asCfgNode() = slash and
+          nodeFrom in [
+              // type-preserving call
+              pathOperand,
+              // data injection
+              dataOperand
+            ]
```

Comment:
Does this not just boil down to the following?
```ql
          slash.getOp() instanceof Div and
          nodeFrom.asCfgNode() = slash.getAnOperand() and
          nodeFrom.getALocalSource() = pathlibPath() and
          nodeTo.asCfgNode() = slash
```
(In which case, more sharing with the `pathlibPath` typetracker might be possible.)

Reply:
Almost, except the argument of type `path lib.Path` does not have to be `nodeFrom`. It still does become much shorter this way, though 👍

Reply:
Sharing is a bit tricky for data injection, because the taint step would prefer to refer to the type tracker to limit the set of nodes.